zebra-state/src/sled_state.rs

```diff
@@ -0,0 +1,195 @@
+            // tx_by_hash: db.open_tree(b"tx_by_hash").unwrap(),
+            // utxo_by_outpoint: db.open_tree(b"utxo_by_outpoint").unwrap(),
+            // sprout_nullifiers: db.open_tree(b"sprout_nullifiers").unwrap(),
+            // sapling_nullifiers: db.open_tree(b"sapling_nullifiers").unwrap(),
+        }
+    }
+
+    /// Queue a finalized block to be committed to the state.
+    pub fn queue(&mut self, queued_block: QueuedBlock) {
+        let prev_hash = queued_block.block.header.previous_block_hash;
+        self.queued_by_prev_hash.insert(prev_hash, queued_block);
+    }
+
+    pub fn process_queue(&mut self) {
+        // Cloning means the closure doesn't hold a borrow of &self,
+        // conflicting with mutable access in the loop below.
+        let hash_by_height = self.hash_by_height.clone();
+        let tip_hash = || {
+            read_tip(&hash_by_height)
+                .expect("inability to look up tip is unrecoverable")
+                .map(|(_height, hash)| hash)
+                .unwrap_or(block::Hash([0; 32]))
+        };
+
+        while let Some(queued_block) = self.queued_by_prev_hash.remove(&tip_hash()) {
+            self.commit_finalized(queued_block)
+        }
+    }
+
+    /// Commit a finalized block to the state. It's the caller's responsibility
+    /// to ensure that blocks are committed in order.
+    fn commit_finalized(&mut self, queued_block: QueuedBlock) {
+        let QueuedBlock { block, rsp_tx } = queued_block;
+
+        // The only valid block without a coinbase height is the genesis
+        // block.  By this point the block has been validated, so if
+        // there's no coinbase height, it must be the genesis block.
+        let height = block.coinbase_height().unwrap_or(block::Height(0));
+        let height_bytes = height.0.to_be_bytes();
+        let hash = block.hash();
+
+        use sled::Transactional;
+        let transaction_result = (
+            &self.hash_by_height,
+            &self.height_by_hash,
+            &self.block_by_height,
+        )
+            .transaction(move |(hash_by_height, height_by_hash, block_by_height)| {
+                // TODO: do serialization above
+                // for some reason this wouldn't move into the closure (??)
+                let block_bytes = block
+                    .zcash_serialize_to_vec()
+                    .expect("zcash_serialize_to_vec has wrong return type");
+
+                // TODO: check highest entry of hash_by_height as in RFC
+
+                hash_by_height.insert(&height_bytes, &hash.0)?;
+                height_by_hash.insert(&hash.0, &height_bytes)?;
+                block_by_height.insert(&height_bytes, block_bytes)?;
+
+                // for some reason type inference fails here
+                Ok::<_, sled::transaction::ConflictableTransactionError>(())
+            });
+
+        let _ = rsp_tx.send(transaction_result.map(|_| hash).map_err(Into::into));
```

Comment:
this seems to work locally

```suggestion
                Ok::<_, sled::transaction::ConflictableTransactionError>(hash)
            });

        let _ = rsp_tx.send(transaction_result.map_err(Into::into));
```

Reply:
Yeah, the thing I was hoping to get rid of is the `::<_, sled::transaction::ConflictableTransactionError>`, which isn't present in the sled examples.

Reply:
that happens because `transaction` is generic over fns of the form `Fn(&Self::View) -> ConflictableTransactionResult<A, E>`. The `E` parameter in this signature is unbound and so you need some way to specify it. If you try doing so by adding `-> ConflictableTransactionResult<_>` however it still fails because it doesn't impl Error anymore. This is because the E parameter in `ConflictableTransactionResult` defaults to `()`, and the `Error` impl on `ConflictableTransactionError` requires that the `T` generic implements `Error`, which the `ConflictableTransactionResult` type alias sets via its `E` parameter to `()` by default. 

`ConflictableTransationError` on the other hand also defaults this parameter but sets it to `sled::Error` by default, so when you use `ConflictableTransactionError` explicitly without specifying the generic parameter it ends up resolving correctly. 

The easiest way to fix this is to make our own type alias that doesn't incorrectly override the T parameter the same way `ConflictableTransactionResult` does.

```rust
type SledResult<T> = Result<T, ConflictableTransactionError>;

// ...

            .transaction(
                move |(hash_by_height, height_by_hash, block_by_height)| -> SledResult<_> {
                    // TODO: do serialization above
                    // for some reason this wouldn't move into the closure (??)
                    let block_bytes = block
                        .zcash_serialize_to_vec()
                        .expect("zcash_serialize_to_vec has wrong return type");

                    // TODO: check highest entry of hash_by_height as in RFC

                    hash_by_height.insert(&height_bytes, &hash.0)?;
                    height_by_hash.insert(&hash.0, &height_bytes)?;
                    block_by_height.insert(&height_bytes, block_bytes)?;

                    Ok(hash)
                },
            );

        let _ = rsp_tx.send(transaction_result.map_err(Into::into));
```

Reply:
Hmm, adding a type ascription seems like a smaller wart to deal with than adding a type alias.

Reply:
the type alias was just to make it cleaner, you could just do `-> Result<_, ConfusableTransactionError>` instead if you prefer